addons/link_tracker/tests/test_mail_render_mixin.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.tests import common
+
+
+class TestMailRenderMixin(common.TransactionCase):
+    def test_shorten_links(self):
+        self.env["mail.render.mixin"]._shorten_links(
+            """
+                <a href="https://github.com" title="title" fake='fake'>test_label</a>
+
+                <a href="https://test_542152qsdqsd.com"/>
+
+                <a href="https://third_test_54212.com">
+                    <img src="imagesrc"/>
+                </a>
+
+                <a
+                    href="https://test_strange_html.com"       title="title"
+                fake='fake'
+                > test_strange_html_label
+                </a>
+
+                <a href="https://test_escaped.com" title="title" fake='fake'> test_escaped &lt; &gt; </a>
+
+                <a href="https://url_with_params.com?a=b&c=d">label</a>
+            """,
+            {},
+        )
+
+        self.assertTrue(
```

Comment:
I think you could simplify the writing and reading of those tests with this global approach

1/ create a list of test candidates
```
links = ["<a href="https://github.com" title="title" fake='fake'>test_label</a>", ...]
```
2/ build the global html
3/ build a list of expected results, something like
```
trackers_to_find = [("MY_URL", "MY_LABEL"), ("https://github.com", "test_label")]
trackers_to_fail = ["MY URL2", "https://test_542152qsdqsd.com"]
```
4/ run the rendering
5/ do assertTrue / False on those list (for tracker in ok: assertTrue(tracker.search()) ...)

That way we can easily add test cases if we find issues, and code will be shorter.

Reply:
Good idea :)